SQLachemy error overriding

neelasha23 opened the pull request:
## Describe your changes

Overriding sqlalchemy error handling for the case:
1. Syntax errors in query (we are using sqlglot to parse the query and prompt possible errors/ suggestions).
2. Password not provided/ in-accessible for postgres connections
3. Incorrect URL connection string for DuckDB

## Issue number

Closes #229 

## Checklist before requesting a review

- [x] Performed a self-review of my code
- [x] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [x] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [x] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)



<!-- readthedocs-preview jupysql start -->
----
:books: Documentation preview :books:: https://jupysql--400.org.readthedocs.build/en/400/

<!-- readthedocs-preview jupysql end -->

neelasha23 commented:
I have added the Slack message link.
Regarding Point 1 

> Show a detailed error message - perhaps even show where the query faulted (analyze it on the fly).

I'm doubtful `where the query faulted` information is available from the underlying DB driver, I couldn't find any relevant resources. It also seems like Sqlite returns all errors as `OperationalError`. (Syntax errors should be returned as ProgrammingError).

Can we do something like :
Detect the most common errors from the error message, e.g., `Syntax error`. If Syntax error then provide a link to this page:
https://sqlite.org/lang.html
Similarly for other DBs.

@edublancas @idomic

tonykploomber commented on src/sql/magic.py:
```diff
@@ -408,11 +408,14 @@ def interactive_execute_wrapper(**kwargs):
                 print(e)
             else:
                 raise
+            print(
```
We can utilize the COMMUNITY String in `ploomber_core/exception.py`, https://github.com/ploomber/core/blob/main/src/ploomber_core/exceptions.py#L12

Probably 
```
from ploomber_core import exceptions
...
print (exceptions. COMMUNITY)
...
```

tonykploomber commented on src/sql/magic.py:
```diff
@@ -408,11 +408,14 @@ def interactive_execute_wrapper(**kwargs):
                 print(e)
             else:
                 raise
+            print(
```
I think what Ido said is correct, let's have test case

neelasha23 replied:
Added tests.

tonykploomber reviewed (changes requested):
minor

neelasha23 commented:
I tried out a few syntactically incorrect queries on sqlglot and these are my inferences:
1. It is able to detect very simple errors like unexpected parantheses, incomplete keyword
2. It detects typos but the error message is not very informative:

```
ParseError: Invalid expression / Unexpected token. Line 1, Col: 7.
INSET INTO foo VALUES (1, 'abc')
```


3. Doesn't detect missing keywords or incomplete queries.
4. Doesn't work on complicated queries.

<img width="1087" alt="Screenshot 2023-04-19 at 12 03 17 AM" src="https://user-images.githubusercontent.com/29096135/232871619-58f7cc7c-c626-4490-9313-9b79a053a282.png">

I'm checking if there are any other libraries

@idomic @tonykploomber

edublancas commented:
We'll not find any library that correctly parses all kinds of SQL dialects; this seems the wrong approach. We might end up with too many false negatives, I believe it's better to look for errors raised by the db driver and show complementary information: either a clearer error message or a link to our docs to fix it.

This issue needs better scoping: what errors do we want to catch, and what do we want to do?

neelasha23 commented:
> This issue needs better scoping: what errors do we want to catch, and what do we want to do?

I was going through stackoverflow posts on SQL errors for different db drivers. There are many errors which are use-case specific , e.g., using SQLAlchemy or Sqlite python client / Errors coming up with specific technologies like Ruby.  Most of these errors are not relevant for our scenario. Majority of the problem still seems to be syntax errors (which the was the starting point of this issue), but there doesn't seem to be any library out there which can detect precisely where the syntax error occurred.
I'm unsure of which errors to include in the scope. Do we have any user data that can give some insights into which errors they are facing frequently?

@edublancas

edublancas commented:
> What if we cover the basics with sqlglot

I don't think we should rely on sqlglot as we've seen that the parser might make mistakes. we don't want to be in a position where we tell the user to "fix their query" when it's a perfectly valid SQL that will run just fine in the database.

a better approach is to let the code run, and only if it fails, proceed with some clear error messages.

neelasha23 commented:
Yeah, so we are not not going to parse query with sqlglot initially. There is a `try..except` block in which the query is being executed. If an Exception raised we show the Slack community link. And only if the error message contains `SyntaxError` and sqloglot is able to detect some error we show it to the user along with the Slack message. Does this look fine?

@idomic  @edublancas

neelasha23 commented on CHANGELOG.md:
```diff
@@ -4,8 +4,8 @@
 * [Doc] Updating connecting guide (by @DaveOkpare) (#56)
-
 * [Feature] Upgrades SQLAlchemy version to 2
+* [Feature] Informative error messages when query execution fails
```
Does this need to be in user guide or in adding in the changelog should suffice?

neelasha23 replied:
@idomic

tonykploomber replied:
@idomic

neelasha23 commented:
@tonykploomber Did you get a chance to review this?

tonykploomber commented on src/sql/error_message.py:
```diff
@@ -0,0 +1,38 @@
+
+def detail(msg, query=None):
+    msg = str(msg)
+    return_msg = "\nLooks like there is some syntax error " "in your query\n"
```
```suggestion
    return_msg = "\nLooks like there is some syntax error " " in your query\n"
```

Also, do we have a test case to cover this?

neelasha23 replied:
yes in `test_error_messages.py` and `test_postgreSQL.py`

tonykploomber replied:
Oh I mean to test the string: `"\nLooks like there is a syntax error " " in your query\n"` in the `capsys.readouterr()`

neelasha23 replied:
No, just added the error specific strings

neelasha23 replied:
Made the changes

tonykploomber reviewed (commented):
Overall LG

neelasha23 commented:
I added `@modify_exceptions` in connections.py file. However, it seems some existing functions were displaying when error in connecting to DB. 

```
PLOOMBER_SUPPORT_LINK_STR = (
    "For technical support: https://ploomber.io/community"
    "\nDocumentation: https://jupysql.ploomber.io/en/latest/connecting.html"
)
``` 

So now `modify_exceptions` msg + the above msg results in slack link displaying twice as below:
```
Can't load plugin: sqlalchemy.dialects:something

To fix it, make sure you are using correct driver name:
Ref: https://docs.sqlalchemy.org/en/20/core/engines.html#database-urls

For technical support: https://ploomber.io/community
Documentation: https://jupysql.ploomber.io/en/latest/connecting.html
If you need help solving this issue, send us a message: https://ploomber.io/community
```

Do we need `PLOOMBER_SUPPORT_LINK_STR` or `modify_exceptions` is fine?

@idomic @edublancas

neelasha23 commented on src/sql/connection.py:
```diff
@@ -200,7 +202,17 @@ def __init__(self, engine, alias=None):
+            detailed_msg = detail(e)
+            if detailed_msg is not None:
+                print(detailed_msg)
+            print(PLOOMBER_SUPPORT_LINK_STR)
```
I used `modify_exceptions` but there is an existing function which is using `PLOOMBER_SUPPORT_LINK_STR`. So `modify_exceptions` + `PLOOMBER_SUPPORT_LINK_STR` is resulting in Community link being displayed twice.

neelasha23 replied:
As of now I only know of `OperationsError` being raised. I'm not sure if any other imp exception might be raised. Should I still change to specific error?

@idomic

neelasha23 replied:
I'll check what changes on removing it.

neelasha23 replied:
Added `modify_exceptions`

neelasha23 commented:
Resolved review comments. But CI keeps failing on Windows
@idomic

neelasha23 commented:
I created a new PR and incrementally added changes from this PR to get to the cause of Windows CI failing. It seems like it is happening due to `UsageError` , but I'm not sure why it is causing test cases to fail on Windows only and not the other platforms.
Here is the commit that caused the cases to start failing: https://github.com/ploomber/jupysql/pull/460/commits/63537bbd19b37c576adb595148e12b8f562cf0e8

Any idea? @edublancas

edublancas commented:
this clean PR is passing: https://github.com/ploomber/jupysql/pull/460 so looks like your changes are breaking something. I looked at the code but I cannot see what happened.

neelasha23 commented:
Here is the commit that caused the cases to start failing: [63537bb](https://github.com/ploomber/jupysql/commit/63537bbd19b37c576adb595148e12b8f562cf0e8)
> 

This is the commit that caused it to break. (I'm guessing it's the `UsageError`. It passes on my local, as well as CI Mac and Ubuntu.

@idomic @edublancas

neelasha23 commented:
CI is passing now
@idomic @tonykploomber

tonykploomber commented on src/sql/error_message.py:
```diff
@@ -0,0 +1,36 @@
+
+def detail(msg, query=None):
+    msg = str(msg)
+    return_msg = "\nLooks like there is a syntax error " "in your query\n"
```
```suggestion
    return_msg = "\nLooks like there is a syntax error " " in your query\n"
```

Add a space between `" "` and `in`

neelasha23 replied:
These quotes are added when formatting using `pkgmt`. The output looks fine in notebook. Do you see any errors?
@tonykploomber

tonykploomber replied:
I see, then no worries

tonykploomber commented on src/tests/test_error_messages.py:
```diff
@@ -0,0 +1,177 @@
+
+
+@pytest.fixture
+def query_description():
```
For me, the naming convention `query_coloumn_with_syntax_error` makes more sense to me.
But I won't block this, up to you.

tonykploomber reviewed (dismissed):
LG
@idomic  Please review too!

yafimvo commented on src/tests/test_error_messages.py:
```diff
@@ -0,0 +1,179 @@
+    assert (
+        "If you need help solving this issue, "
+        "send us a message: https://ploomber.io/community" in str(err)
+    )
```
I think we can put these 2 messages in a global variable since they might change in the future and fail our tests.

neelasha23 replied:
Done. fetching from ploomber_core now

yafimvo reviewed (changes requested):
lg. minor comment.

While I was running a few tests, I got this error which I couldn't reproduce.

```
*  duckdb://
(duckdb.InvalidInputException) Invalid Input Error: Attempting to execute an unsuccessful or closed pending query result
Error: Invalid Error: Current transaction is aborted (please ROLLBACK)
[SQL: select * from penguins.csv]
(Background on this error at: https://sqlalche.me/e/20/f405)
```

It looks like we need to restart the kernel otherwise nothing will work.

neelasha23 commented:
> While I was running a few tests, I got this error which I couldn't reproduce.
> 
> ```
> *  duckdb://
> (duckdb.InvalidInputException) Invalid Input Error: Attempting to execute an unsuccessful or closed pending query result
> Error: Invalid Error: Current transaction is aborted (please ROLLBACK)
> [SQL: select * from penguins.csv]
> (Background on this error at: https://sqlalche.me/e/20/f405)
> ```
> 
> It looks like we need to restart the kernel otherwise nothing will work.

Do you remember the sequence of queries?

@yafimvo

neelasha23 commented on src/sql/error_message.py:
```diff
@@ -0,0 +1,38 @@
+    msg = str(msg)
+    return_msg = SYNTAX_ERROR
+    if "syntax error" in msg:
+        query_list = query.split(";")
```
usually it's at the end of query. can u give an example of `;` in between query?

neelasha23 replied:
It should split this into 2 different queries which is what we want. because we need to parse each query through sqlglot

neelasha23 replied:
We decided to parse using sqlglot only if there is an exception (otherwise there's no other better way to detect the syntax error). We wont be parsing using sqlglot initially to prevent false cases.

https://github.com/ploomber/jupysql/pull/400#issuecomment-1515621581

neelasha23 replied:
So if there's an exception we do this:
1. User may write multiple `;` separated queries, so split and extract each query separately.
2. Parse each query through sqlglot and append this the error message.
3. Finally raise `UsageError` with the above message

neelasha23 replied:
https://github.com/ploomber/jupysql/pull/400#issuecomment-1515038679

neelasha23 commented on src/tests/test_magic.py:
```diff
@@ -647,14 +651,14 @@ def test_error_on_invalid_connection_string(ip_empty, clean_conns):
 
 
-invalid_connection_string_format = """\
+invalid_connection_string_format = f"""\
```
this is an existing one. I'll check and revert

neelasha23 replied:
`invalid_connection_string_existing_conns` : Runs an extra step `ip_empty.run_cell("%sql sqlite://")` before running an invalid connection. That's why these are different cases. Only error message is same

neelasha23 replied:
Removed `invalid_connection_string_existing_conns` message string and used `invalid_connection_string_format`

neelasha23 commented:
> 1. We Found a few issues with Duckdb connection so we probably want to cover it.

Added an exception handling when passing incorrect URL strings. This will look like:

<img width="1135" alt="Screenshot 2023-05-07 at 1 40 41 AM" src="https://user-images.githubusercontent.com/29096135/236644894-835f84da-24bc-4520-a50c-188e73ecb486.png">

(Original error from driver + custom ploomber links)
Does this look ok?

> 2. We can also cover a missing column error.

This would involve parsing the query and extracting the table names and column names. Also, missing column names can be used for any clause `SELECT`, `GROUP BY` etc.. so we need to cover all these cases and add tests. I think it would be better to handle this in another issue.

Updated scope of issue:
1. Handling syntax errors and showing more defined error messages (like pointing to the incorrect part of the query/ providing alternate suggestions)
2. Password not provided when connecting to PostgreSQL using URL formats.
3. Error message when invalid duckdb connection string

@idomic @edublancas @yafimvo @tonykploomber

yafimvo commented:
> Added an exception handling when passing incorrect URL strings. This will look like:

I think it looks OK. Do we know if it's a bad URL issue? If so, I would consider removing the original error from the driver since it may confuse the user (he ran `%sql duckdb://invalid` and the error is referring to `connect()`)



> This would involve parsing the query and extracting the table names and column names. Also, missing column names can be used for any clause `SELECT`, `GROUP BY` etc.. so we need to cover all these cases and add tests. I think it would be better to handle this in another issue.

I agree 

@neelasha23 @idomic

neelasha23 commented:
> I think it looks OK. Do we know if it's a bad URL issue? If so, I would consider removing the original error from the driver since it may confuse the user (he ran `%sql duckdb://invalid` and the error is referring to `connect()`)

No we are trying to connect to the engine and if it throws any exception we raise `UsageError` with custom Ploomber links.
The original message gives some context `host=invalid`.

Any thoughts on point 3 [here](https://github.com/ploomber/jupysql/pull/400#issuecomment-1535137826) for syntax errors

yafimvo commented:
> Any thoughts on point 3 [here](https://github.com/ploomber/jupysql/pull/400#issuecomment-1535137826) for syntax errors

What if we change the order? Our error first, then the driver error, and then the reference to the Slack channel?

neelasha23 commented:
> What if we change the order? Our error first, then the driver error, and then the reference to the Slack channel?

Can be done . Maybe we should add a statement `Original driver error` before the driver error?
@idomic @yafimvo

yafimvo commented:
> Can be done . Maybe we should add a statement `Original driver error` before the driver error?

Yes, I think it's good and makes the error message more informative. Distinguishing between the errors will allow us to do something like this in the future.

https://github.com/ploomber/jupysql/assets/103026689/40b436fb-f54f-44e0-b2d1-61da8fa4b524

@neelasha23 @idomic

neelasha23 commented:
> Yes, I think it's good and makes the error message more informative. Distinguishing between the errors will allow us to do something like this in the future.
> 

Yes, , nice one.

I have added the message for now and it looks like:
<img width="1094" alt="Screenshot 2023-05-09 at 4 05 21 PM" src="https://github.com/ploomber/jupysql/assets/29096135/9fdc6c28-a8bc-4190-b5bb-ce5df0bb3ba3">

@yafimvo @idomic

yafimvo commented:
> I have added the message for now and it looks like:

To keep it consistent with the connection error, can we include it in the `UsageError` under the same structure?

UsageError: {{Our message}}

{{Suggestions}}

{{Driver error}}

{{slack and docs references}}

neelasha23 commented:
> To keep it consistent with the connection error, can we include it in the `UsageError` under the same structure?
> 

Can me done. Let me refactor the code.

neelasha23 commented:
Fixed the error messages as per the discussion with @yafimvo  above.

Case1: `sqlglot` unable to detect any errors/ suggestions
<img width="1065" alt="Screenshot 2023-05-09 at 6 38 54 PM" src="https://github.com/ploomber/jupysql/assets/29096135/7610fcf2-1dce-4ec3-943f-e5a1b3b4999a">

Case2: Syntax error position detected
<img width="1070" alt="Screenshot 2023-05-09 at 6 38 46 PM" src="https://github.com/ploomber/jupysql/assets/29096135/799dc8af-d72b-4507-8533-abed7d565716">


Case3: Alternate query suggestion
<img width="1073" alt="Screenshot 2023-05-09 at 6 38 29 PM" src="https://github.com/ploomber/jupysql/assets/29096135/ffe99c5e-d11f-4a18-99ee-1579b28badeb">


Is this fine? Any other changes required?
@idomic

neelasha23 commented:
> Yes that looks fine, please fix the conflict and we're good to go.

Done
@idomic

tonykploomber commented:
Good to me

tonykploomber commented:
@idomic 
Is anything you want to review before we merge this one?
